ggml/src/ggml-amx/mmq.cpp

```diff
@@ -0,0 +1,2584 @@
+      // out features is 32x
+      ne0 % (TILE_N * 2) == 0;
+}
+
+// NB: mixed dtype gemm with Advanced Matrix Extensions (Intel AMX)
+//
+// src0: weight in shape of {N, K}, quantized
+// src1: input  in shape of {M, K}, float32
+// dst:  output in shape of {M, N}, float32
+//
+// the function performs: dst = src1 @ src0.T
+//
+void ggml_mul_mat_amx(struct ggml_tensor * dst, int nth, int ith, void * wdata, int wsize) {
+  struct ggml_tensor * src0 = dst->src[0];
+  struct ggml_tensor * src1 = dst->src[1];
+
+  const enum ggml_type TYPE = src0->type;
+
+  // f16 only has avx512 kernels for now,
+  // amx kernels will be added once 6th gen xeon is released.
+  const bool is_floating_type = TYPE == GGML_TYPE_F16;
+
+  const int M = dst->ne[1];
+  const int N = dst->ne[0];
+  const int K = src0->ne[0];
+  const int ldc = dst->nb[1] / dst->nb[0];
+
+  if (is_floating_type) {
+    constexpr int BLOCK_M = 4;
+    constexpr int BLOCK_N = 6;
+    const int MB = div_up(M, BLOCK_M);
+    const int NB = div_up(N, BLOCK_N);
+
+    parallel_for(nth, ith, MB * NB, [&](int begin, int end) {
+      GGML_DISPATCH_FLOATING_TYPES(TYPE, [&] {
+        for (int i = begin; i < end; ++i) {
+          int mb = i / NB;
+          int nb = i % NB;
+
+          int mb_start = mb * BLOCK_M;
+          int mb_size = std::min(BLOCK_M, M - mb_start);
+          int nb_start = nb * BLOCK_N;
+          int nb_size = std::min(BLOCK_N, N - nb_start);
+
+          switch (mb_size << 4 | nb_size) {
+            case 0x12: LAUNCH_TINYGEMM_KERNEL_AVX(1, 2); break;
+            case 0x14: LAUNCH_TINYGEMM_KERNEL_AVX(1, 4); break;
+            case 0x16: LAUNCH_TINYGEMM_KERNEL_AVX(1, 6); break;
+            case 0x22: LAUNCH_TINYGEMM_KERNEL_AVX(2, 2); break;
+            case 0x24: LAUNCH_TINYGEMM_KERNEL_AVX(2, 4); break;
+            case 0x26: LAUNCH_TINYGEMM_KERNEL_AVX(2, 6); break;
+            case 0x32: LAUNCH_TINYGEMM_KERNEL_AVX(3, 2); break;
+            case 0x34: LAUNCH_TINYGEMM_KERNEL_AVX(3, 4); break;
+            case 0x36: LAUNCH_TINYGEMM_KERNEL_AVX(3, 6); break;
+            case 0x42: LAUNCH_TINYGEMM_KERNEL_AVX(4, 2); break;
+            case 0x44: LAUNCH_TINYGEMM_KERNEL_AVX(4, 4); break;
+            case 0x46: LAUNCH_TINYGEMM_KERNEL_AVX(4, 6); break;
+            default: fprintf(stderr, "Unexpected block size!\n");
+          }
+        }
+      });
+    });
+    return;
+  }
+
+  #pragma omp single
+  {
+    GGML_DISPATCH_QTYPES(TYPE, [&] {
+      const size_t row_size_A = K / blck_size * sizeof(vec_dot_type);
+      GGML_ASSERT(wsize >= int(M * row_size_A));
+
+      // Q4_0, Q4_1, Q8_0 handles 1 TILE_K per blck_size
+      // Q4_K, Q5_K, Q6_K, IQ4_XS handles 8 TILE_K per blck_size
+      GGML_ASSERT(TILE_K == blck_size || TILE_K * 8 == blck_size);
+      // pack mat B to vnni format
+      if (src0->extra == nullptr) {
+        const size_t row_size_B = get_row_size<type, blck_size>(K);
+        src0->extra = aligned_alloc(64, N * row_size_B);
```

Comment:
When in this memory freed?

Reply:
right now this memory is not freed until program exits. After I updated the prepacking logic and several rounds of rebasing, the old free logic no longer works.

@slaren does cpu has a mempool structure which allows dynamic alloc from the mempool? Any suggestions?

Reply:
The CPU backend does not have any mempool that could be used to allocate this memory. But even if it did, this approach would double the memory requirements for the model weights, which I don't think is an acceptable tradeoff in most cases. It is also likely to cause issues when the tensors are not static, such as when using KV cache quantization.

There is a way to implement this kind of tensor data layout transformations with ggml-backend. In short, it is possible to create a type of `ggml_backend_buffer` that performs these transformations transparently when accessing the tensor data through the `set_tensor`/`get_tensor` interfaces. The assumption would be that all tensors allocated in a buffer of this type would use the different layout. Before performing the matrix multiplication, you could check the type of buffer where the tensor is allocated to determine if AMX can be used. This would also require extensive changes to the llama.cpp model loading code to use this buffer type. Alternatively, the existing CPU buffer type could be modified to perform these transformations, but that would also require changing other operations of the CPU backend to be able to use the different layout.

In the past this also has been implemented by adding new tensor data types with the modified layout (eg. `Q4_0_4_4` and others for ARM), but I think it would be hard to justify the maintenance cost of adding new quantization types.

My conclusion is that there isn't a simple solution to implement this, and it would need to be done along with significant refactoring of other parts of the code (eg. the llama.cpp model loading code). llama.cpp supports offloading large matrix multiplication to a GPU, which is usually a lot faster than even very optimized CPU code, so I find it hard to justify to spend too much effort into this, but of course contributions are welcome.

Reply:
@slaren is it possible to do an inplace replacement in the original tensor with the AMX required layout when AMX can be used? This way would be very easy to implement, not too much code change is needed. But I am not sure if there is any risk of doing this?

Reply:
The most significant risk would be breaking the semantics of the way tensors are supposed to work in ggml. For example, after loading a model into tensors and evaluating it, an application may want to read the tensor data to serialize the model back to disk. The tensor data is not hidden from applications, so that is expected to work. Another consequence would be that the tensor data would no longer be in the format that other operations expect. For instance, in models where the token embeddings and output layer use the same tensor, this would break the `get_rows` operation.

Reply:
changing amx as a ggml-backend to do memory management.

Reply:
@slaren could you pleas help review this one again? I have updated the implementation with ggml-backend.